Split DifferentiableGraphOp out of graph_executor.cpp

zasdfgbnm opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28856 Allow printing DifferentiableGraphBackward objects
* #28626 Expose DifferentiableGraphBackward
* **#28563 Split DifferentiableGraphOp out of graph_executor.cpp**

DifferentiableGraphOp and the graph executor are two independent things.
Spliting them out will make future development more convenient.

This PR is purely copy and paste, and could be easily reviewed.

In the future I will further refactor things in backward.cpp,
in order to provide Python API to visualize backward graph created by
autodiff (#28549).

zasdfgbnm committed:
Split DifferentiableGraphOp out of graph_executor.cpp

DifferentiableGraphOp and the graph executor are two independent things.
Spliting them out will make future development more convenient.

This PR is purely copy and paste, and could be easily reviewed.

In the future I will further refactor things in backward.cpp,
in order to provide Python API to visualize backward graph created by
autodiff (#28549).

[ghstack-poisoned]

zasdfgbnm committed:
Update on "Split DifferentiableGraphOp out of graph_executor.cpp"


DifferentiableGraphOp and the graph executor are two independent things.
Spliting them out will make future development more convenient.

This PR is purely copy and paste, and could be easily reviewed.

In the future I will further refactor things in backward.cpp,
in order to provide Python API to visualize backward graph created by
autodiff (#28549).

[ghstack-poisoned]

zasdfgbnm commented:
Besides the copy-paste, I also made some tiny changes to fix the clang-tidy.

ZolotukhinM commented on caffe2/CMakeLists.txt:
```diff
@@ -363,6 +363,7 @@ if (NOT INTERN_BUILD_MOBILE OR NOT BUILD_CAFFE2_MOBILE)
     ${TORCH_SRC_DIR}/csrc/jit/pickler.cpp
     ${TORCH_SRC_DIR}/csrc/jit/unpickler.cpp
     ${TORCH_SRC_DIR}/csrc/jit/graph_executor.cpp
+    ${TORCH_SRC_DIR}/csrc/jit/backward.cpp
```
When we add a new file, we also need to add it to `tools/build_variables.py` (that file generates build files for internal builds).

Also, a nit: the name `backward.cpp` seems to me too generic, maybe use something more concrete? `grad_executor` for instance, but I don't insist on that particular option.

zasdfgbnm committed:
Update on "Split DifferentiableGraphOp out of graph_executor.cpp"


DifferentiableGraphOp and the graph executor are two independent things.
Spliting them out will make future development more convenient.

This PR is purely copy and paste, and could be easily reviewed.

In the future I will further refactor things in backward.cpp,
in order to provide Python API to visualize backward graph created by
autodiff (#28549).

[ghstack-poisoned]

zasdfgbnm commented:
@ZolotukhinM `grad_executor` sounds good to me and I have updated the name. I also updated `tools/build_variables.py`.

ZolotukhinM commented:
> @ZolotukhinM `grad_executor` sounds good to me and I have updated the name. I also updated `tools/build_variables.py`.

Thanks! I'm not very well familiar with how differentiable graphs are designed now, so I'd wait for other people to take a look at the actual change.

zdevito reviewed (commented):
This refactor looks good. I haven't looked at the followups yet. It has a minor conflict with https://github.com/pytorch/pytorch/pull/28535 which is a big PR that is difficult to land (and should land today). So it would be great if we could wait a day before landing this one.

zasdfgbnm commented:
I will keep an eye on #28535 and rebase this once #28535 get landed.

Krovatkin commented:
@zasdfgbnm My PR landed, if you could wait maybe until late afternoon to make sure that nothing blows up before rebasing your PR, that would be ideal in case I will need to revert some of my code.

zasdfgbnm commented:
@Krovatkin Thanks for reminding.

zasdfgbnm committed:
Update on "Split DifferentiableGraphOp out of graph_executor.cpp"


DifferentiableGraphOp and the graph executor are two independent things.
Spliting them out will make future development more convenient.

This PR is purely copy and paste, and could be easily reviewed.

In the future I will further refactor things in backward.cpp,
in order to provide Python API to visualize backward graph created by
autodiff (#28549).

[ghstack-poisoned]

zasdfgbnm committed:
Update on "Split DifferentiableGraphOp out of graph_executor.cpp"


DifferentiableGraphOp and the graph executor are two independent things.
Spliting them out will make future development more convenient.

This PR is purely copy and paste, and could be easily reviewed.

In the future I will further refactor things in backward.cpp,
in order to provide Python API to visualize backward graph created by
autodiff (#28549).

[ghstack-poisoned]

zasdfgbnm commented:
I guess this could be landed now.
